src/protocols/secure_channel/CASESession.cpp

```diff
@@ -224,12 +233,29 @@ class CASESession::WorkHelper
         bool cancel = false;
         // Execute callback in background thread; data must be OK with this
         helper->mStatus = helper->mWorkCallback(helper->mData, cancel);
+
+        // Check the status and call OnWorkFailedCallback
+        if (helper->mStatus != CHIP_NO_ERROR)
+        {
+            if (auto * session = helper->mSession.load())
+            {
+                (session->*(helper->mOnWorkFailedCallback))(helper->mStatus);
+            }
+            return;
+        }
+
         VerifyOrReturn(!cancel && !helper->IsCancelled());
         // Hold strong ptr while work is outstanding
         helper->mStrongPtr.swap(strongPtr);
         auto status = DeviceLayer::PlatformMgr().ScheduleWork(AfterWorkHandler, reinterpret_cast<intptr_t>(helper));
+
         if (status != CHIP_NO_ERROR)
         {
+            // Scheduling AfterWorkHandler failed, lets call OnWorkFailedCallback
+            if (auto * session = helper->mSession.load())
+            {
+                (session->*(helper->mOnWorkFailedCallback))(helper->mStatus);
```

Comment:
How do we know it is safe to call the callback on the session object from a background thread? If WorkHelper requires the error handling callback to be safe to call from background thread (like WorkCallback), then it needs to be document that requirement (like WorkCallback).

Reply:
Added a comment, PTAL!